File: .claude/skills/agent-skill-kit-commons/packages/mjs/src/deps/inversion.mjs
/**
 * Tier-rank table. A skill at rank R may depend only on skills at rank <= R.
 * Spec tiers: org / team / app / project.
 * Extended tiers (only meaningful when --extra-tiers is on): company, enterprise, application.
 */
export const TIER_RANK = {
  org: 0,
  team: 1,
  app: 2,
  project: 3,
  // extended
  company: -2,
  enterprise: -1,
  application: 2,
};

export function checkInversions(registry) {
  const inversions = [];
  for (const [name, deps] of registry.edges) {
    const fromNode = registry.nodes.get(name);
    if (!fromNode || !fromNode.tier) continue;
    const fromRank = TIER_RANK[fromNode.tier];
    if (fromRank === undefined) continue;
    for (const dep of deps) {
      if (dep.origin !== "in_repo") continue;
      const toNode = registry.nodes.get(dep.name);
      if (!toNode || !toNode.tier) continue;
      const toRank = TIER_RANK[toNode.tier];
      if (toRank === undefined) continue;
      if (toRank > fromRank) {
        inversions.push({
          from: name,
          to: dep.name,
          fromTier: fromNode.tier,
          toTier: toNode.tier,
        });
      }
    }
  }
  return inversions;
}

File: .claude/skills/agent-skill-kit-commons/packages/py/src/skills_ref/deps/inversion.py
from __future__ import annotations

TIER_RANK = {
    "org": 0,
    "team": 1,
    "app": 2,
    "project": 3,
    # extended
    "company": -2,
    "enterprise": -1,
    "application": 2,
}


def check_inversions(registry) -> list:
    inversions = []
    for name, deps in registry.edges.items():
        from_node = registry.nodes.get(name)
        if not from_node or not from_node.get("tier"):
            continue
        from_rank = TIER_RANK.get(from_node["tier"])
        if from_rank is None:
            continue
        for dep in deps:
            if dep.origin != "in_repo":
                continue
            to_node = registry.nodes.get(dep.name)
            if not to_node or not to_node.get("tier"):
                continue
            to_rank = TIER_RANK.get(to_node["tier"])
            if to_rank is None:
                continue
            if to_rank > from_rank:
                inversions.append({
                    "from": name,
                    "to": dep.name,
                    "fromTier": from_node["tier"],
                    "toTier": to_node["tier"],
                })
    return inversions
